### PR TITLE
feat(recovery): add recovery key metrics

### PR DESCRIPTION
**Pull request by vbudhram**
Fixes https://github.com/mozilla/fxa-auth-server/issues/2538

Below is a snippet of flow events for creating and consuming recovery keys. I believe it covers all of the metrics in https://github.com/mozilla/fxa-auth-server/issues/2538.

```
3|content- | {"event":"flow.settings.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":524,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:13.212Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":533,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:13.221Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":18721,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:31.409Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.refresh","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":22710,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:35.398Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.clicked-support-link","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":25052,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:37.740Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-password.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":33853,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:46.541Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-password.engage","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":35369,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:48.057Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-password.submit","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":40624,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:53.312Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-password.invalidPassword","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":40816,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:49:53.504Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-password.submit","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":49342,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:02.030Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-password.success","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":49790,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:02.478Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.recovery-key.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":49837,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:02.525Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.recovery-key.print-key","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":51688,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:04.376Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.recovery-key.download-key","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":54883,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:07.571Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":57325,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:10.013Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-revoke.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":60295,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:12.983Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-revoke.submit","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":61617,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:14.305Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.confirm-revoke.success","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":61645,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:14.333Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
3|content- | {"event":"flow.settings.account-recovery.view","flow_id":"5d139de0bd6d03d313b116878e164ecc6b8f2a2928d5fa7e612a819d4448a922","flow_time":61663,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:50:14.351Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
```

```
3|content- | {"event":"flow.account-recovery-confirm-key.view","flow_id":"84d8318ea0ec654f82544e1db1860315036a6a98e1a353337f3f776a8f510106","flow_time":124049,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:54:34.560Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-confirm-key.engage","flow_id":"84d8318ea0ec654f82544e1db1860315036a6a98e1a353337f3f776a8f510106","flow_time":128836,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:54:39.347Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-confirm-key.submit","flow_id":"84d8318ea0ec654f82544e1db1860315036a6a98e1a353337f3f776a8f510106","flow_time":131438,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:54:41.949Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-confirm-key.invalidRecoveryKey","flow_id":"84d8318ea0ec654f82544e1db1860315036a6a98e1a353337f3f776a8f510106","flow_time":131679,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":87740,"time":"2018-08-14T18:54:42.190Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-confirm-key.submit","flow_id":"2baa860968964dc15d342bf69cc29d0ef236bd4e2744263970a21c6c2010e48d","flow_time":45963,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":95570,"time":"2018-08-14T19:09:25.662Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-confirm-key.success","flow_id":"5b8402b33b6255a80965a7ad6d5317fa0768c39674f7e73fa6e6efd6c3fd9512","flow_time":34505,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":95570,"time":"2018-08-14T19:22:47.302Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-reset-password.view","flow_id":"2baa860968964dc15d342bf69cc29d0ef236bd4e2744263970a21c6c2010e48d","flow_time":46193,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":95570,"time":"2018-08-14T19:09:25.892Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-reset-password.engage","flow_id":"2baa860968964dc15d342bf69cc29d0ef236bd4e2744263970a21c6c2010e48d","flow_time":47076,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":95570,"time":"2018-08-14T19:09:26.775Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.account-recovery-reset-password.submit","flow_id":"2baa860968964dc15d342bf69cc29d0ef236bd4e2744263970a21c6c2010e48d","flow_time":51660,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":95570,"time":"2018-08-14T19:09:31.359Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
3|content- | {"event":"flow.reset-password-with-recovery-key-verified.view","flow_id":"2baa860968964dc15d342bf69cc29d0ef236bd4e2744263970a21c6c2010e48d","flow_time":51939,"hostname":"vijaybudhram-ms.local","locale":"en","op":"flowEvent","pid":95570,"time":"2018-08-14T19:09:31.638Z","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:58.0) Gecko/20100101 Firefox/58.0","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync","utm_campaign":"fx-forgot-password","utm_content":"fx-reset-password","utm_medium":"email"}
```

### CODE REVIEW

**Comment by shane-tomlinson** [comment]
@vbudhram - most of these errors will only go to datadog, whereas I imagine you are going to want flow events to send to Amplitude and Re:dash.

File: app/scripts/lib/verification-reasons.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ define(function (require, exports, module) {
   return {
     FORCE_AUTH: 'force_auth',
     PASSWORD_RESET: 'reset_password',
+    PASSWORD_RESET_WITH_RECOVERY_KEY: 'reset_password_with_recovery_key',
```
Comment (vbudhram) [inline app/scripts/lib/verification-reasons.js]: Added a specific view for when a user completes resetting their password with recovery key. This was used to differentiate it in metrics from a regular reset.

**Comment by vbudhram** [comment]
@mozilla/fxa-devs r?

**Comment by shane-tomlinson** [comment]
>  most of these errors will only go to datadog, whereas I imagine you are going to want flow events to send to Amplitude and Re:dash.

Have I lost touch with how flow events work? I'm going to look at this again tomorrow.

**Comment by shane-tomlinson** [comment]
> Have I lost touch with how flow events work? I'm going to look at this again tomorrow.

OMG. Never code review when tired. I read "logFlowEvent" as "logViewEvent". It's all clear now.

**Comment by shane-tomlinson** [comment]
@vbudhram - looks like this needs rebased.

**Comment by vbudhram** [comment]
@shane-tomlinson Thanks for review! I added the cancel metrics logging. I don't believe the metric names were very important as long as they were known, ref https://github.com/mozilla/fxa-auth-server/issues/2538#issuecomment-407431091.

**Pull request merged**